packages/react-client/src/hooks/devices.ts

```diff
@@ -1,33 +1,98 @@
-  const devices = state.devices.camera.devices ?? [];
-  const activeDevice = state.devices.camera.deviceInfo;
+  const devices = deviceState.devices ?? [];
+  const activeDevice = deviceState.media?.deviceInfo ?? null;
 
   return { ...trackManager, streamedTrack, streamedTrackId, stream, devices, activeDevice };
 }
 
 export function useMicrophone(): AudioDevice {
   const {
-    state,
     audioTrackManager: { currentTrack, ...trackManager },
   } = useFishjamContext();
 
+  const { deviceState } = useAudioDeviceManager();
+
   const isAudioPlaying = currentTrack?.vadStatus === "speech";
   const streamedTrackId = currentTrack?.trackId ?? null;
   const streamedTrack = currentTrack?.track ?? null;
   const stream = currentTrack?.stream ?? null;
-  const devices = state.devices.microphone.devices ?? [];
-  const activeDevice = state.devices.microphone.deviceInfo;
+  const devices = deviceState.devices ?? [];
+  const activeDevice = deviceState.media?.deviceInfo ?? null;
 
   return { ...trackManager, streamedTrack, streamedTrackId, isAudioPlaying, stream, devices, activeDevice };
 }
+
+type InitializeDevicesProps = { autoInitialize?: boolean };
+
+export const useInitializeDevices = (props?: InitializeDevicesProps) => {
+  const areDevicesInitialized = useRef(false);
+  const { videoDeviceManagerRef, audioDeviceManagerRef } = useFishjamContext();
+
+  const initializeDevices = useCallback(async () => {
+    if (areDevicesInitialized.current) return;
+    const videoManager = videoDeviceManagerRef.current;
+    const audioManager = audioDeviceManagerRef.current;
+
+    const constraints = {
+      video: videoManager.getConstraints(true),
+      audio: audioManager.getConstraints(true),
+    };
+
+    const previousDevices = {
+      video: videoManager.getLastDevice(),
+      audio: audioManager.getLastDevice(),
+    };
+
+    let [stream, deviceErrors] = await getAvailableMedia(constraints);
+
+    const devices = await navigator.mediaDevices.enumerateDevices();
+
+    const videoDevices = devices.filter(({ kind }) => kind === "videoinput");
+    const audioDevices = devices.filter(({ kind }) => kind === "audioinput");
+
+    if (stream) {
+      [stream, deviceErrors] = await getCorrectedResult(stream, deviceErrors, devices, constraints, previousDevices);
+    }
+
+    videoManager.initialize(
+      stream,
+      stream?.getVideoTracks()?.[0] ?? null,
+      videoDevices,
+      !!constraints.video,
+      deviceErrors.video,
+    );
+    audioManager.initialize(
+      stream,
+      stream?.getAudioTracks()?.[0] ?? null,
+      audioDevices,
+      !!constraints.audio,
+      deviceErrors.audio,
+    );
+    areDevicesInitialized.current = true;
+  }, [videoDeviceManagerRef, audioDeviceManagerRef]);
+
+  const autoInitialize = Boolean(props?.autoInitialize);
+
+  useEffect(() => {
+    if (!autoInitialize) return;
+    initializeDevices();
+  }, [initializeDevices, autoInitialize]);
+
+  return { initializeDevices };
```

Comment:
**Future improvement idea**: WDYT about adding additional state like `initializing` that will inform the user that the manager has started, so the user can disable buttons or display some kind of popup like "Grant access to camera and microphone"?

Of course in separate PR

Reply:
Sound like a plan! :)

Reply:
https://linear.app/swmansion/issue/FCE-456/return-device-manager-status-from-useinitializedevices